core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -43,6 +43,15 @@ import scala.util.Try
 object KafkaController extends Logging {
   val InitialControllerEpoch = 1
   val InitialControllerEpochZkVersion = 1
+
+  /**
+   * ControllerEventThread will shutdown once it sees this event
+   */
+  private[controller] case object ShutdownEventThread extends ControllerEvent {
+    def state = ControllerState.ControllerChange
```

Comment:
Perhaps we should use a new state like ControllerShutdown?

Reply:
I've done this now. I was unsure if we wanted to add a new state for this so tried to reuse one that was closest, but it seems reasonable to add a more specific one.